working/2364 - primary constructors/feature-specification.md

```diff
@@ -170,15 +172,20 @@ class", but it is consistent with the rest of the language to say that this
 particular primary constructor is a "constant constructor". Hence `class
 const Name` rather than `const class Name`.
 
-The modifier `final` on a parameter in a primary constructor has no meaning
-for the parameter itself, because there is no scope where the parameter can
-be accessed. Hence, this modifier is used to specify that the instance
-variable declared by this primary constructor parameter is `final`.
+The modifier `final` on a parameter in a primary constructor has the usual
+effect that the parameter itself cannot be modified. The only location
+where such modifications can occur is in an assertion, and they "should"
+not have side effects, so we can basically ignore this. However, much more
+importantly, this modifier is also used to specify that the instance
+variable declaration which is induced by this primary constructor parameter
+is `final`.
 
 In the case where the constructor is constant, and in the case where the
 declaration is an `extension type` or an `enum` declaration, the modifier
 `final` on every instance variable is required. Hence, it can be omitted
-from the formal parameter in the primary constructor:
+from the formal parameter in the primary constructor, because it is implied
```

Comment:
Fwiw this does make converting a constructor from const to non-const potentially error prone - you could easily forget to add `final` to all the fields that you want to keep final.

Reply:
True. But how often does it happen that a constant constructor get turned into a non-constant one? After all, it's going to break every constant expression out there where this constructor is invoked.

Reply:
I have no idea, I assume it would only happen when a class that used to be const no longer can be.

It begs the question though, is there really much value in allowing omitting `final` versus just requiring the user to write it? Is it more readable with or without the explicit `final`?

Reply:
I'll preserve the ability (it was always an option, not a requirement) to omit `final` when every instance variable in the declaration must be final (that is, the primary constructor is `const` or the declaration is an enum or an extension type). There are two reasons: (1) I'd like to maintain the emphasis on brevity, and _requiring_ `final` or `var` on every formal parameter that introduces an instance variable will make the syntax considerably more verbose. (2) Extension types already use this abbreviation, and it would be nice if we can say that the `(...)` in the header of an extension type declaration is just a primary constructor, with no special exceptions.

If everybody insists that we must have `var` or `final` on every formal parameter of a primary constructor that introduces an instance variable then it shouldn't be hard to create the syntax and rules for that, but right now I want to keep the more concise form in mind.